### PR TITLE
Tests for libtinyalsa

### PR DESCRIPTION
**Pull request by dvdli**
The total coverages of these tests are 46.2% for lines and 63.5% for functions. The lines and functions for plugin feature are counted as well, but we don't have any plugin implementation. Therefore, no test tests that part of code.

### CODE REVIEW

File: tests/src/pcm_in_test.cc [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        .start_threshold = 0,
+        .stop_threshold = 0,
+        .silence_threshold = 0,
+        .silence_size = 0,
+    };
+
+    pcm* pcm_object;
+};
+
+TEST_F(PcmInTest, GetDealy) {
```
Comment (tay10r) [inline tests/src/pcm_in_test.cc:75]: Minor typo here

File: tests/include/pcm_test_device.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+#ifndef TEST_LOOPBACK_CAPTURE_DEVICE
+#define TEST_LOOPBACK_CAPTURE_DEVICE 1
+#endif
+
+constexpr unsigned int kLoopbackCard = TEST_LOOPBACK_CARD;
+constexpr unsigned int kLoopbackPlaybackDevice = TEST_LOOPBACK_PLAYBACK_DEVICE;
+constexpr unsigned int kLoopbackCaptureDevice = TEST_LOOPBACK_CAPTURE_DEVICE;
+
+} // namespace testing
+} // namespace tinyalse
```
Comment (tay10r) [inline tests/include/pcm_test_device.h:52]: Minor typo here and in all the following trailing namespace comments in the other test files.

File: tests/src/mixer_test.cc [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+        if (mixer_ctl_get_type(control) == MIXER_CTL_TYPE_INT) {
+            unsigned int number_of_values = mixer_ctl_get_num_values(controls[i]);
+            std::unique_ptr<long []> values = std::make_unique<long []>(number_of_values);
+            mixer_ctl_get_array(control, values.get(), number_of_values);
+            for (unsigned int value_id = 0; value_id < number_of_values; ++value_id) {
+                int max = mixer_ctl_get_range_max(control);
+                int min = mixer_ctl_get_range_min(control);
+                int value = values[value_id];
+                int percent = mixer_ctl_get_percent(control, value_id);
+                if (mixer_ctl_set_percent(const_cast<mixer_ctl *>(control), value_id, 100) == 0) {
```
Comment (gkasten) [inline tests/src/mixer_test.cc:240]: Consider using a symbol for 100

File: tests/src/mixer_test.cc [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+
+    if (control == nullptr) {
+        GTEST_SKIP() << "No volume control was found in the controls list.";
+    }
+
+    auto *local_mixer_object = mixer_object;
+    int percent = mixer_ctl_get_percent(control, 0);
+    std::thread thread([local_mixer_object, control, percent] () {
+        std::this_thread::sleep_for(std::chrono::milliseconds(50));
+        mixer_ctl_set_percent(const_cast<mixer_ctl *>(control), 0, percent == 100 ? 0 : 100);
```
Comment (gkasten) [inline tests/src/mixer_test.cc:279]: same

**Review by gkasten — COMMENTED** [review]
overall

**Comment by tay10r** [comment]
This is a really great addition and long overdue. I've only done one look through of the tests so far but I didn't notice any testing of error conditions. Just as one example, testing that a PCM opened with a bad config will:

 1. Return a PCM that is not null
 2. Contain an error message string that is not null or empty (`err_msg != NULL` and `err_msg[0] != 0`)

I would not technically consider this part of the review, but it may be worth considering in the future.

File: tests/src/pcm_loopback_test.cc [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+
+    std::thread capture([pcm_in, &stopping] {
+        size_t buffer_size = pcm_frames_to_bytes(pcm_in, kDefaultPeriodSize);
+        unsigned int frames = pcm_bytes_to_frames(pcm_in, buffer_size);
+        auto buffer = std::make_unique<unsigned char[]>(buffer_size);
+        int32_t counter = 0;
+        while (!stopping) {
+            EXPECT_EQ(pcm_readi(pcm_in, buffer.get(), frames), frames);
+            counter++;
+            if (counter >= 5) {
```
Comment (gkasten) [inline tests/src/pcm_loopback_test.cc:188]: Can you please add comment to explain why it is 5 ?

File: tests/src/pcm_params_test.cc [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+#include <gtest/gtest.h>
+
+#include "tinyalsa/pcm.h"
+
+namespace tinyalsa {
+namespace testing {
+
+static inline unsigned int OrAllBits(const pcm_mask *mask) {
+    unsigned int res = 0;
+    for (uint32_t i = 0; i < 32 / sizeof(unsigned int); ++i) {
```
Comment (gkasten) [inline tests/src/pcm_params_test.cc:44]: Where does magic number 32 come from?

**Review by gkasten — APPROVED** [review]
Looks good, thank you

**Comment by dvdli** [comment]
Thanks for your reviewing. Actually, there are some cases returning null pointers when to open fails. In the pcm_open_by_name function, it returns a null pointer when the format of name is not matched. Therefore, I want to fix them, and then add tests for testing the behaviors of those cases.

I also found some minor bugs (and include the issue #184 ) when I was implementing these tests. I will start to fix them and documents as well as add tests for them.

> This is a really great addition and long overdue. I've only done one look through of the tests so far but I didn't notice any testing of handling bad arguments. Just as one example, testing that a PCM opened with a bad config will:
> 
> 1. Return a PCM that is not null
> 2. Contain an error message string that is not null or empty (`err_msg != NULL` and `err_msg[0] != 0`)
> 
> I would not technically consider this part of the review, but it may be worth considering in the future.

File: tests/src/pcm_in_test.cc (hunk 1 above)
Comment (dvdli) [reply tests/src/pcm_in_test.cc]: Done

File: tests/include/pcm_test_device.h (hunk 2 above)
Comment (dvdli) [reply tests/include/pcm_test_device.h]: Done

File: tests/src/pcm_loopback_test.cc (hunk 5 above)
Comment (dvdli) [reply tests/src/pcm_loopback_test.cc:198]: Done.

File: tests/src/mixer_test.cc (hunk 3 above)
Comment (dvdli) [reply tests/src/mixer_test.cc]: Done

File: tests/src/mixer_test.cc (hunk 4 above)
Comment (dvdli) [reply tests/src/mixer_test.cc]: Done

File: tests/src/pcm_params_test.cc (hunk 6 above)
Comment (dvdli) [reply tests/src/pcm_params_test.cc]: The total bytes of mask is 32. I add a constant to describe.

**Pull request merged**